### PR TITLE
EPE-148 - Fix EOF problem in datastream

### PR DESCRIPTION
**Pull request by jamesfolk-blockone**
This is in reference to this: https://blockone.atlassian.net/browse/EPE-148

* I have added defensive programming practices to cfile to ensure that a user cannot use it if it is closed.

### CODE REVIEW

File: include/fc/io/cfile.hpp [hunk 1]
```diff
@@ -27,14 +27,16 @@ class cfile_datastream;
...
    cfile()
-     : _file(nullptr, &fclose)
+     : _open(false),
+       _file_path(""),
+       _file(nullptr, &fclose)
    {}
 
-   void set_file_path( fc::path file_path ) {
+   void set_file_path( const fc::path &file_path ) {
       _file_path = std::move( file_path );
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: The `std::move` here should be removed if you are going to pass by `const&`. Why was this changed?

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: yes you are right. i'll fix it.

File: include/fc/io/cfile.hpp [hunk 2]
```diff
@@ -27,14 +27,16 @@ class cfile_datastream;
...
 public:
    cfile()
-     : _file(nullptr, &fclose)
+     : _open(false),
+       _file_path(""),
+       _file(nullptr, &fclose)
    {}
 
-   void set_file_path( fc::path file_path ) {
+   void set_file_path( const fc::path &file_path ) {
```
Comment (rusty-block-one) [inline include/fc/io/cfile.hpp]: The std::move of a const& results in a copy of an lvalue. This is no different from stating "_file_path = file_path". It may also confuse the reader into thinking rvalue move semantics are being implemented.

File: include/fc/io/cfile.hpp (hunk 1 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: it was an attempt to remove the function from calling the copy constructor unnecessarily. I reverted it. since the implementation a move, maybe an r-value param might be better?

File: include/fc/io/cfile.hpp [hunk 3]
```diff
@@ -49,6 +51,9 @@ class cfile {
    ///         "ab+" - open for binary update - create if does not exist
    ///         "rb+" - open for binary update - file must exist
    void open( const char* mode ) {
+       if(_open)
```
Comment (rusty-block-one) [inline include/fc/io/cfile.hpp]: For correctness, I would suggest setting _open back to false.

**Review by rusty-block-one — CHANGES_REQUESTED** [review]

File: include/fc/io/cfile.hpp (hunk 3 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: good call.

**Review by rusty-block-one — APPROVED** [review]

File: include/fc/io/cfile.hpp [hunk 4]
```diff
@@ -27,14 +27,21 @@ class cfile_datastream;
...
+     : _open(false),
+       _file_path(""),
+       _file(nullptr, &fclose)
    {}
 
+   ~cfile() {
+       if(_open)
+           fclose(_file.get());
+   }
+
```
Comment (heifner) [inline include/fc/io/cfile.hpp:36]: indentation seems to be 4 spaces instead of 3 spaces as used throughout rest of code. Can you please fix.

File: include/fc/io/cfile.hpp [hunk 5]
```diff
@@ -27,14 +27,21 @@ class cfile_datastream;
...
    cfile()
-     : _file(nullptr, &fclose)
+     : _open(false),
+       _file_path(""),
+       _file(nullptr, &fclose)
    {}
 
+   ~cfile() {
+       if(_open)
+           fclose(_file.get());
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Doesn't this call `fclose` twice on `_file`?
Seems like you could just get rid of the `unique_file` and do it manually since you are anyway.

File: include/fc/io/cfile.hpp [hunk 6]
```diff
@@ -65,77 +80,118 @@ class cfile {
...
-   int getc() { 
-      int ret = fgetc(_file.get());  
-      if (ret == EOF) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+        if(feof(_file.get()))return true;
+
+        int c = fgetc(_file.get());
+        if(EOF == c)return true;
+
+        return (EOF == ungetc(c, _file.get()));
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Is this complicated version fo `eof()` needed? Can you provide some information about why these changes are needed. I'm a bit concerned about the `ungetc`:
"A call to fseek, fsetpos or rewind on stream will discard any characters previously put back into it with this function."

File: include/fc/io/cfile.hpp [hunk 7]
```diff
@@ -27,14 +27,21 @@ class cfile_datastream;
...
+       if(_open)
+           fclose(_file.get());
+   }
+
    void set_file_path( fc::path file_path ) {
       _file_path = std::move( file_path );
    }
 
-   fc::path get_file_path() const {
+   const fc::path &get_file_path() const {
```
Comment (brianjohnson5972) [inline include/fc/io/cfile.hpp]: the pattern we are suppose to follow is "const fc::path& " (don't know if Larry has actually released the coding style document yet, but that is the preferred way)

**Review by heifner — CHANGES_REQUESTED** [review]

File: include/fc/io/cfile.hpp (hunk 5 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: yes. i'm second guessing myself-

File: include/fc/io/cfile.hpp (hunk 6 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: i was being paranoid. I removed it.

File: include/fc/io/cfile.hpp [hunk 8]
```diff
@@ -65,77 +80,118 @@ class cfile {
    }
 
    void seek( long loc ) {
-      if( 0 != fseek( _file.get(), loc, SEEK_SET ) ) {
+       if(!_open)
+           throw std::ios_base::failure("cfile is not open");
+
+       if( 0 != fseek( _file.get(), loc, SEEK_SET ) ) {
```
Comment (brianjohnson5972) [inline include/fc/io/cfile.hpp]: The initial problem talked about EOF issues and had a fix to this method.  Was that determined to not be essential, or was it actually because the file wasn't open?  I would think the open check just prevents getting a confusing error from fseek.

File: include/fc/io/cfile.hpp [hunk 9]
```diff
@@ -65,77 +80,118 @@ class cfile {
...
-   bool eof() const { return feof(_file.get()) != 0; }
+    bool eof() const  {
+        if(!_open)
+            throw std::ios_base::failure("cfile is not open");
 
-   int getc() { 
-      int ret = fgetc(_file.get());  
-      if (ret == EOF) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+        if(feof(_file.get()))return true;
```
Comment (brianjohnson5972) [inline include/fc/io/cfile.hpp]: move the return to the next line

File: include/fc/io/cfile.hpp [hunk 10]
```diff
@@ -65,77 +80,118 @@ class cfile {
...
+
+        int c = fgetc(_file.get());
+        if(EOF == c)return true;
```
Comment (brianjohnson5972) [inline include/fc/io/cfile.hpp]: move the return to the next line

**Review by brianjohnson5972 — DISMISSED** [review]

File: include/fc/io/cfile.hpp (hunk 8 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: I just removed it.

File: include/fc/io/cfile.hpp [hunk 11]
```diff
@@ -49,6 +51,11 @@ class cfile {
    ///         "ab+" - open for binary update - create if does not exist
    ///         "rb+" - open for binary update - file must exist
    void open( const char* mode ) {
+      if(is_open()) {
+         fclose(_file.get());
+         _open = false;
+      }
+
       _file.reset( FC_FOPEN( _file_path.generic_string().c_str(), mode ) );
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: `_file` is closed twice here. Once above and again with `reset`.  Seems like we either should just go with `_open` flag and raw `fclose` OR remove `_open` flag and just go with `!= nullptr` on `unique_file`.

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: thanks for pointing this out.
it seems that using specifying the delete process for the unique pointer on a FILE, might not be the best of ideas.
Maybe just removing the delete process of the unique pointer might be the better idea. I'll make a PR to show you what i mean-

File: include/fc/io/cfile.hpp [hunk 12]
```diff
@@ -25,16 +25,27 @@ class cfile_datastream;
  * std::ios_base::failure exception thrown for errors.
  */
 class cfile {
+private:
```
Comment (jamesfolk-blockone) [inline include/fc/io/cfile.hpp]: this code block will ensure that there is only one way to ctor a cfile

File: include/fc/io/cfile.hpp [hunk 13]
```diff
@@ -15,7 +15,7 @@
 namespace fc {
 
 namespace detail {
-   using unique_file = std::unique_ptr<FILE, decltype( &fclose )>;
+   using unique_file = std::unique_ptr<FILE>;
```
Comment (swatanabe-b1) [inline include/fc/io/cfile.hpp]: This is definitely wrong.  You can't `delete` a `FILE*`, which is what the vanilla `unique_ptr` will try to do.

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: right- it seems to me that the unique_ptr is only complicating things.

**Comment by swatanabe-b1** [comment]
I'm not sure what the point of this PR is.  It doesn't address the issue in EPE-148 at all.  Instead it's just changing the usage of unique_ptr which originally worked fine.

**Comment by jamesfolk-blockone** [comment]
> I'm not sure what the point of this PR is. It doesn't address the issue in EPE-148 at all. Instead it's just changing the usage of unique_ptr which originally worked fine.

The problem I am trying to fix is a hard to track down EOF issue with cfile. 
With the previous implementation of cfile, when you called the open for the for  the first  time, fclose is called on an undefined  instance of FILE* (unique_ptr::reset())

**Comment by swatanabe-b1** [comment]
> The problem I am trying to fix is a hard to track down EOF issue with cfile.
> With the previous implementation of cfile, when you called the open for the for the first time, fclose is called on an undefined instance of FILE* (unique_ptr::reset())

It isn't undefined, it's null and `fclose` is not called: "if and only if the old value of the stored pointer, old_p, was not equal to nullptr, calls get_deleter()(old_p)" [unique.ptr.single.modifiers].

**Comment by swatanabe-b1** [comment]
The issue with EOF is that `remaining()` may return `true` even if the file position indicator is at the end of the file, because it uses `feof`, but that isn't what `feof` means.

**Comment by jamesfolk-blockone** [comment]
> > The problem I am trying to fix is a hard to track down EOF issue with cfile.
> > With the previous implementation of cfile, when you called the open for the for the first time, fclose is called on an undefined instance of FILE* (unique_ptr::reset())
> 
> It isn't undefined, it's null and `fclose` is not called: "if and only if the old value of the stored pointer, old_p, was not equal to nullptr, calls get_deleter()(old_p)" [unique.ptr.single.modifiers].

I see. i did  a quick reference check and that is true-

**Comment by jamesfolk-blockone** [comment]
> The issue with EOF is that `remaining()` may return `true` even if the file position indicator is at the end of the file, because it uses `feof`, but that isn't what `feof` means.

Thank you for that. Is that written somewhere? if so, I must have glazed over it.

**Comment by swatanabe-b1** [comment]
It was a comment in the first PR linked in the issue.

**Comment by jamesfolk-blockone** [comment]
> It was a comment in the first PR linked in the issue.

I see. Is this the comment from June 8th 2020 from this link? https://github.com/EOSIO/eos/pull/9192

**Comment by swatanabe-b1** [comment]
That is a different, but related issue.  in_avail is another function that is close, but not quite what we need.  The one I was referring to is https://github.com/EOSIO/eos/pull/9183#discussion_r435921222.

**Comment by jamesfolk-blockone** [comment]
> That is a different, but related issue. in_avail is another function that is close, but not quite what we need. The one I was referring to is [EOSIO/eos#9183 (comment)](https://github.com/EOSIO/eos/pull/9183#discussion_r435921222).

remaining is a very strange function. is it needed?

**Comment by swatanabe-b1** [comment]
> remaining is a very strange function. is it needed?

Delete it and see what doesn't compile?

**Comment by swatanabe-b1** [comment]
Also, I don't really see it as strange.  The intended behavior is pretty simple: are there any characters left that can be read?

**Comment by jamesfolk-blockone** [comment]
> Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?

The reasoning as to why I find it strange is because you could essentially use eof for that use case, instead of a specialized eof wrapped function which has to be maintained.

**Comment by jamesfolk-blockone** [comment]
> Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?

it built when i commented out the remaining function

**Comment by brianjohnson5972** [comment]
> > Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?
> 
> it built when i commented out the remaining function

You will need to make sure that all the repos that use it still build, not just EOSIO/eos

**Comment by jamesfolk-blockone** [comment]
> > > Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?
> > 
> > 
> > it built when i commented out the remaining function
> 
> You will need to make sure that all the repos that use it still build, not just EOSIO/eos

Ok, Thank you Brian. That is good to know. it sounds easier if i fix the `remaining` function.

**Comment by jamesfolk-blockone** [comment]
> > > > Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?
> > > 
> > > 
> > > it built when i commented out the remaining function
> > 
> > 
> > You will need to make sure that all the repos that use it still build, not just EOSIO/eos
> 
> Ok, Thank you Brian. That is good to know. it sounds easier if i fix the `remaining` function.

I believe I have fixed the `remaining` function.

File: include/fc/io/cfile.hpp [hunk 14]
```diff
@@ -188,13 +227,21 @@ class datastream<fc::cfile, void> : public fc::cfile {
 
    bool get(char& c) {
       c = this->getc();
+      _current_c = c;
```
Comment (swatanabe-b1) [inline include/fc/io/cfile.hpp]: `get` is not the only function that can affect `remaining`.

File: include/fc/io/cfile.hpp [hunk 15]
```diff
@@ -188,13 +227,21 @@ class datastream<fc::cfile, void> : public fc::cfile {
...
 
    fc::cfile&       storage() { return *this; }
    const fc::cfile& storage() const { return *this; }
+private:
+   datastream(const datastream &) = delete;
+   datastream(const datastream &&) = delete;
+   const datastream& operator=(const datastream &) = delete;
+   const datastream&& operator=(const datastream &&) = delete;
+
+   char _current_c = EOF;
```
Comment (swatanabe-b1) [inline include/fc/io/cfile.hpp]: EOF is an `int`, not a `char`

File: include/fc/io/cfile.hpp [hunk 16]
```diff
@@ -188,13 +227,21 @@ class datastream<fc::cfile, void> : public fc::cfile {
 
    bool get(char& c) {
       c = this->getc();
+      _current_c = c;
       return true;
    }
 
-   bool remaining() { return !this->eof(); }
+   bool remaining() { return (EOF != _current_c); }
```
Comment (swatanabe-b1) [inline include/fc/io/cfile.hpp]: This still has essentially the same problem as the original.  It checks whether you attempted to read past the end, not whether the datastream is at the end.

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: Thank you  for  pointing that out. I addressed this issue,  now.

File: include/fc/io/cfile.hpp [hunk 17]
```diff
@@ -53,10 +53,26 @@ class cfile {
       if( !_file ) {
          throw std::ios_base::failure( "cfile unable to open: " +  _file_path.generic_string() + " in mode: " + std::string( mode ) );
       }
+
+      fseek(_file.get(), 0L, SEEK_END);
+      _size = ftell(_file.get());
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: What if you write to the file after you open it?

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: good catch. thanks.

File: include/fc/io/cfile.hpp (hunk 15 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: EOF is defined as `#define	EOF	(-1)`

File: include/fc/io/cfile.hpp [hunk 18]
```diff
@@ -94,14 +119,24 @@ class cfile {
...
+      if(!is_open())
+         throw std::ios_base::failure("cfile is not open");
+
+      size_t result = fwrite( d, WRITE_SIZE, n, _file.get() );
       if( result != n ) {
          throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
                                        " unable to write " + std::to_string( n ) + " bytes; only wrote " + std::to_string( result ) );
       }
+      // Keep up with the size  of the file.
+      _size +=  (WRITE_SIZE * n);
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Do we have a restriction that only one cfile can read/write to a file? I don't think there is that restriction before this change.

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: so what you are saying  is that it  might have been possible for  two instances of cfile to open the same file at the  same time, before these  changes?

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp:133]: so what you are saying  is that it  might have been possible for  two instances of cfile to open the same file at the  same time, before these  changes?

Comment (heifner) [reply include/fc/io/cfile.hpp]: Yes. I don't think there was anything to prevent that use case.

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: ok. that is a good  point. Is this a requirement? Who do I talk to about  this?

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: maybe remaining is  only a requirement for `cfile_datastream`?

Comment (heifner) [reply include/fc/io/cfile.hpp]: Can't you use the impl you had? 
https://www.geeksforgeeks.org/eof-and-feof-in-c/
https://www.tutorialspoint.com/c_standard_library/c_function_feof.htm

```
    bool eof() const  {
        if(!_open)
            throw std::ios_base::failure("cfile is not open");

        if(feof(_file.get()))return true;

        int c = fgetc(_file.get());
        if(EOF == c)return true;

        return (EOF == ungetc(c, _file.get()));
    }
```

Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: good point.

File: include/fc/io/cfile.hpp [hunk 19]
```diff
@@ -53,10 +53,25 @@ class cfile {
...
       _open = true;
    }
 
+   bool remaining(){
+      if(!is_open() || eof())
+         return false;
+
+      int c(getc());
+      bool isEOF(eof());
+      ungetc(c, _file.get());
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Check return code for error.

File: include/fc/io/cfile.hpp [hunk 20]
```diff
@@ -53,10 +53,25 @@ class cfile {
...
+
       _open = true;
    }
 
+   bool remaining(){
+      if(!is_open() || eof())
+         return false;
+
+      int c(getc());
+      bool isEOF(eof());
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Really don't like this syntax. Can you at least use `{}` instead of `()` for both of these.

File: include/fc/io/cfile.hpp [hunk 21]
```diff
@@ -94,14 +118,20 @@ class cfile {
    }
 
    void write( const char* d, size_t n ) {
-      size_t result = fwrite( d, 1, n, _file.get() );
+      if(!is_open())
+         throw std::ios_base::failure("cfile is not open");
+
+      size_t result = fwrite( d, WRITE_SIZE, n, _file.get() );
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Where and what is WRITE_SIZE? This needs to be one because `n` is in terms of char.

File: include/fc/io/cfile.hpp [hunk 22]
```diff
@@ -141,6 +182,11 @@ class cfile {
    cfile_datastream create_datastream();
 
 private:
+   cfile(const cfile &) = delete;
+   cfile(const cfile &&) = delete;
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Should be `cfile( cfile&& ) = delete;`

File: include/fc/io/cfile.hpp [hunk 23]
```diff
@@ -141,6 +182,11 @@ class cfile {
...
+   const cfile& operator=(const cfile &) = delete;
+   const cfile&& operator=(const cfile &&) = delete;
```
Comment (heifner) [inline include/fc/io/cfile.hpp]: Should be `cfile& operator=( cfile&& ) = delete;`

File: include/fc/io/cfile.hpp (hunk 22 above)
Comment (heifner) [reply include/fc/io/cfile.hpp]: Same for other types below.

File: include/fc/io/cfile.hpp (hunk 23 above)
Comment (heifner) [reply include/fc/io/cfile.hpp]: Same for others below.

**Review by heifner — CHANGES_REQUESTED** [review]

File: include/fc/io/cfile.hpp (hunk 20 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: no problem. I like to use the initialization list even though it is non-traditional. I will just use the traditional syntax..

File: include/fc/io/cfile.hpp (hunk 21 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: oops- thanks for catching that. i  cleaned up the code with out trying to compile

File: include/fc/io/cfile.hpp (hunk 22 above)
Comment (jamesfolk-blockone) [reply include/fc/io/cfile.hpp]: lesson learned*

**Review by heifner — APPROVED** [review]
